Exception with hidden cause during annotation processing

### Expected Behavior
When an exception is thrown during annotation processing, it should include the underlying cause (the original exception) so that the developer can fix the issue.

### Current Behavior
An exception is caught in AnnotationProcessingCompileTask.instantiateProcessor and the code then throws an IllegalArgumentException without including the original caught exception as the cause. This makes it impossible to identify the cause of the exception.

The AnnotationProcessingCompileTask.loadProcessor method does the same thing.

### Context
I had not included all dependencies needed by the annotation processor on the annotation processor classpath. The underlying cause was a ClassNotFoundException, however I couldn't see that until I edited the source for AnnotationProcessingCompileTask.

So the ClassNotFoundException was caught in AnnotationProcessingCompileTask.instantiateProcessor and not included as the cause of the IllegalArgumentException thrown from this method.

### Proposed changes
Simply pass the original exception as the 2nd parameter to the new IllegalArgumentException throw in both AnnotationProcessingCompileTask's instantiateProcessor and loadProcessor methods. I've confirmed this fixes the issue and I don't imagine this will have any side-effects.

### Stack trace
The stack trace should've looked like this but was missing all the "Caused by" lines.

Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileGenerateJava'.
Caused by: java.lang.IllegalArgumentException: Could not instantiate annotation processor 'com.example.SomeAnnotationProcessor'
	at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.instantiateProcessor(AnnotationProcessingCompileTask.java:150)
Caused by: java.lang.reflect.InvocationTargetException
Caused by: java.lang.NoClassDefFoundError: com/example/ExampleAnnotation
Caused by: java.lang.ClassNotFoundException: com.example.ExampleAnnotation

--- a/subprojects/docs/src/docs/release/notes.md
+++ b/subprojects/docs/src/docs/release/notes.md
@@ -8,6 +8,7 @@ We would like to thank the following community contributors to this release of G
 [Daiki Hirabayashi](https://github.com/dhirabayashi),
 [Sebastian Schuberth](https://github.com/sschuberth),
 [Frieder Bluemle](https://github.com/friederbluemle),
+[Brick Tamland](https://github.com/mleveill),
 and [Stefan Oehme](https://github.com/oehme).
 
 <!-- 

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java
@@ -58,7 +58,7 @@ class AnnotationProcessingCompileTask implements JavaCompiler.CompilationTask {
     private final List<File> annotationProcessorPath;
     private final AnnotationProcessingResult result;
 
-    private URLClassLoader processorClassloader;
+    private ClassLoader processorClassloader;
     private boolean called;
 
     AnnotationProcessingCompileTask(JavaCompiler.CompilationTask delegate, Set<AnnotationProcessorDeclaration> processorDeclarations, List<File> annotationProcessorPath, AnnotationProcessingResult result) {
@@ -117,7 +117,7 @@ class AnnotationProcessingCompileTask implements JavaCompiler.CompilationTask {
         delegate.setProcessors(processors);
     }
 
-    private URLClassLoader createProcessorClassLoader() {
+    ClassLoader createProcessorClassLoader() {
         return new URLClassLoader(
             DefaultClassPath.of(annotationProcessorPath).getAsURLArray(),
             getFilteredClassLoader(delegate.getClass().getClassLoader())
@@ -128,7 +128,7 @@ class AnnotationProcessingCompileTask implements JavaCompiler.CompilationTask {
         try {
             return processorClassloader.loadClass(declaredProcessor.getClassName());
         } catch (ClassNotFoundException e) {
-            throw new IllegalArgumentException("Annotation processor '" + declaredProcessor.getClassName() + "' not found");
+            throw new IllegalArgumentException("Annotation processor '" + declaredProcessor.getClassName() + "' not found", e);
         }
     }
 
@@ -136,7 +136,7 @@ class AnnotationProcessingCompileTask implements JavaCompiler.CompilationTask {
         try {
             return (Processor) processorClass.getConstructor().newInstance();
         } catch (Exception e) {
-            throw new IllegalArgumentException("Could not instantiate annotation processor '" + processorClass.getName() + "'");
+            throw new IllegalArgumentException("Could not instantiate annotation processor '" + processorClass.getName() + "'", e);
         }
     }
 

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTask.java
@@ -32,13 +32,14 @@ import org.gradle.internal.concurrent.CompositeStoppable;
 import javax.annotation.processing.Processor;
 import javax.tools.JavaCompiler;
 import java.io.File;
+import java.lang.reflect.InvocationTargetException;
 import java.net.URLClassLoader;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
 import java.util.Set;
 
-import static org.gradle.api.internal.tasks.compile.filter.AnnotationProcessorFilter.*;
+import static org.gradle.api.internal.tasks.compile.filter.AnnotationProcessorFilter.getFilteredClassLoader;
 
 /**
  * Wraps another {@link JavaCompiler.CompilationTask} and sets up its annotation processors
@@ -128,7 +129,7 @@ class AnnotationProcessingCompileTask implements JavaCompiler.CompilationTask {
         try {
             return processorClassloader.loadClass(declaredProcessor.getClassName());
         } catch (ClassNotFoundException e) {
-            throw new IllegalArgumentException("Annotation processor '" + declaredProcessor.getClassName() + "' not found", e);
+            throw new IllegalArgumentException("Annotation processor '" + declaredProcessor.getClassName() + "' not found", unwrapCause(e));
         }
     }
 
@@ -136,8 +137,15 @@ class AnnotationProcessingCompileTask implements JavaCompiler.CompilationTask {
         try {
             return (Processor) processorClass.getConstructor().newInstance();
         } catch (Exception e) {
-            throw new IllegalArgumentException("Could not instantiate annotation processor '" + processorClass.getName() + "'", e);
+            throw new IllegalArgumentException("Could not instantiate annotation processor '" + processorClass.getName() + "'", unwrapCause(e));
         }
+    }
+
+    private Throwable unwrapCause(Throwable throwable) {
+        if (throwable instanceof InvocationTargetException) {
+            return throwable.getCause();
+        }
+        return throwable;
     }
 
     private Processor decorateForIncrementalProcessing(Processor processor, IncrementalAnnotationProcessorType type, AnnotationProcessorResult processorResult) {

--- a/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
+++ b/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
@@ -1,0 +1,144 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile
+
+import org.gradle.api.internal.tasks.compile.incremental.processing.AnnotationProcessingResult
+import org.gradle.api.internal.tasks.compile.incremental.processing.IncrementalAnnotationProcessorType
+import org.gradle.api.internal.tasks.compile.processing.AnnotationProcessorDeclaration
+import org.hamcrest.BaseMatcher
+import spock.lang.Issue
+import spock.lang.Specification
+import static spock.util.matcher.HamcrestSupport.that
+
+import java.lang.annotation.ElementType
+import java.lang.annotation.Target
+import javax.annotation.processing.AbstractProcessor
+import javax.annotation.processing.RoundEnvironment
+import javax.annotation.processing.SupportedAnnotationTypes
+import javax.annotation.processing.SupportedSourceVersion
+import javax.lang.model.SourceVersion
+import javax.lang.model.element.TypeElement
+import javax.tools.JavaCompiler
+
+class AnnotationProcessingCompileTaskTest extends Specification {
+
+    List<File> procPath = Collections.emptyList()
+    AnnotationProcessingResult result = Stub(AnnotationProcessingResult)
+    JavaCompiler.CompilationTask compileTask = Stub(JavaCompiler.CompilationTask)
+    ClassLoader noopLoader = Stub(ClassLoader)
+
+    @Issue("gradle/gradle#8996")
+    def "stacktrace contains root cause"() {
+        given:
+            def proc = new AnnotationProcessorDeclaration(processorClassName, IncrementalAnnotationProcessorType.UNKNOWN)
+            def processTask = createProcessingTask(proc)
+
+        when:
+            processTask.call()
+
+        then:
+            def e = thrown(Throwable)
+            def root = getRootCause(e)
+            root.message.equals(errorMsg)
+            that root.stackTrace, hasFrame(throwingClass)
+
+        where:
+            processorClassName              | throwingClass                      | errorMsg
+            FaultyProcessor.class.getName() | FaultyProcessor.class.getName()    | "Whoops!"
+            "you.wont.find.Me"              | LimitedClassLoader.class.getName() | "Unexpected class load: you.wont.find.Me"
+    }
+
+    def createProcessingTask(AnnotationProcessorDeclaration proc) {
+        Set<AnnotationProcessorDeclaration> procs = Collections.singletonList(proc)
+        return new AnnotationProcessingCompileTask(compileTask, procs, procPath, result) {
+            ClassLoader createProcessorClassLoader() {
+                return new LimitedClassLoader(noopLoader)
+            }
+        }
+    }
+
+    def getRootCause(Throwable t) {
+        if (t.getCause() == null || t.getCause() == t) {
+            return t
+        }
+        return getRootCause(t.getCause())
+    }
+
+    private hasFrame(final String className) {
+        [
+            matches: { frames ->
+                def foundFrame = null
+
+                for (StackTraceElement frame : frames) {
+                    if (className.equals(frame.className)) {
+                        foundFrame = frame
+                    }
+                }
+
+                if (foundFrame == null) {
+                    return false
+                }
+
+                assert foundFrame.fileName != null
+                assert foundFrame.lineNumber > 0
+                assert foundFrame.methodName != null
+                return true
+            },
+
+            describeTo: { description ->
+                description.appendText("stack trace contain frame for class ${className}")
+            },
+
+            describeMismatch: { frames, description ->
+                description.appendValue(frames.toArrayString()).appendText(" did not have frame for class ${className}")
+            }
+        ] as BaseMatcher
+    }
+}
+
+class LimitedClassLoader extends ClassLoader {
+    LimitedClassLoader(ClassLoader parent) {
+        super(parent)
+    }
+
+    @Override
+    public Class<?> loadClass(String name) throws ClassNotFoundException {
+        if (FaultyProcessor.class.getName().equals(name)) {
+            return FaultyProcessor.class
+        } else {
+            throw new ClassNotFoundException("Unexpected class load: " + name)
+        }
+    }
+}
+
+@Target(ElementType.TYPE)
+@interface RunsFaultyProcessor {
+    String QUALIFIED_NAME = "org.gradle.api.internal.tasks.compile.RunsFaultyProcessor"
+}
+
+@SupportedAnnotationTypes(RunsFaultyProcessor.QUALIFIED_NAME)
+@SupportedSourceVersion(SourceVersion.RELEASE_6)
+class FaultyProcessor extends AbstractProcessor {
+    public FaultyProcessor() {
+        throw new RuntimeException("Whoops!")
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment env) {
+        return false
+    }
+}

--- a/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
+++ b/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
@@ -22,10 +22,7 @@ import org.gradle.api.internal.tasks.compile.processing.AnnotationProcessorDecla
 import org.hamcrest.BaseMatcher
 import spock.lang.Issue
 import spock.lang.Specification
-import static spock.util.matcher.HamcrestSupport.that
 
-import java.lang.annotation.ElementType
-import java.lang.annotation.Target
 import javax.annotation.processing.AbstractProcessor
 import javax.annotation.processing.RoundEnvironment
 import javax.annotation.processing.SupportedAnnotationTypes
@@ -33,6 +30,10 @@ import javax.annotation.processing.SupportedSourceVersion
 import javax.lang.model.SourceVersion
 import javax.lang.model.element.TypeElement
 import javax.tools.JavaCompiler
+import java.lang.annotation.ElementType
+import java.lang.annotation.Target
+
+import static spock.util.matcher.HamcrestSupport.that
 
 class AnnotationProcessingCompileTaskTest extends Specification {
 
@@ -44,22 +45,22 @@ class AnnotationProcessingCompileTaskTest extends Specification {
     @Issue("gradle/gradle#8996")
     def "stacktrace contains root cause"() {
         given:
-            def proc = new AnnotationProcessorDeclaration(processorClassName, IncrementalAnnotationProcessorType.UNKNOWN)
-            def processTask = createProcessingTask(proc)
+        def proc = new AnnotationProcessorDeclaration(processorClassName, IncrementalAnnotationProcessorType.UNKNOWN)
+        def processTask = createProcessingTask(proc)
 
         when:
-            processTask.call()
+        processTask.call()
 
         then:
-            def e = thrown(Throwable)
-            def root = getRootCause(e)
-            root.message.equals(errorMsg)
-            that root.stackTrace, hasFrame(throwingClass)
+        def e = thrown(Throwable)
+        def root = getRootCause(e)
+        root.message == errorMsg
+        that root.stackTrace, hasFrame(throwingClass)
 
         where:
-            processorClassName              | throwingClass                      | errorMsg
-            FaultyProcessor.class.getName() | FaultyProcessor.class.getName()    | "Whoops!"
-            "you.wont.find.Me"              | LimitedClassLoader.class.getName() | "Unexpected class load: you.wont.find.Me"
+        processorClassName              | throwingClass                      | errorMsg
+        FaultyProcessor.class.getName() | FaultyProcessor.class.getName()    | "Whoops!"
+        "you.wont.find.Me"              | LimitedClassLoader.class.getName() | "Unexpected class load: you.wont.find.Me"
     }
 
     def createProcessingTask(AnnotationProcessorDeclaration proc) {
@@ -116,8 +117,8 @@ class LimitedClassLoader extends ClassLoader {
     }
 
     @Override
-    public Class<?> loadClass(String name) throws ClassNotFoundException {
-        if (FaultyProcessor.class.getName().equals(name)) {
+    Class<?> loadClass(String name) throws ClassNotFoundException {
+        if (FaultyProcessor.class.getName() == name) {
             return FaultyProcessor.class
         } else {
             throw new ClassNotFoundException("Unexpected class load: " + name)

--- a/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
+++ b/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
@@ -16,10 +16,10 @@
 
 package org.gradle.api.internal.tasks.compile
 
+import com.google.common.base.Throwables
 import org.gradle.api.internal.tasks.compile.incremental.processing.AnnotationProcessingResult
 import org.gradle.api.internal.tasks.compile.incremental.processing.IncrementalAnnotationProcessorType
 import org.gradle.api.internal.tasks.compile.processing.AnnotationProcessorDeclaration
-import org.hamcrest.BaseMatcher
 import spock.lang.Issue
 import spock.lang.Specification
 
@@ -33,14 +33,7 @@ import javax.tools.JavaCompiler
 import java.lang.annotation.ElementType
 import java.lang.annotation.Target
 
-import static spock.util.matcher.HamcrestSupport.that
-
 class AnnotationProcessingCompileTaskTest extends Specification {
-
-    List<File> procPath = Collections.emptyList()
-    AnnotationProcessingResult result = Stub(AnnotationProcessingResult)
-    JavaCompiler.CompilationTask compileTask = Stub(JavaCompiler.CompilationTask)
-    ClassLoader noopLoader = Stub(ClassLoader)
 
     @Issue("gradle/gradle#8996")
     def "stacktrace contains root cause"() {
@@ -53,9 +46,9 @@ class AnnotationProcessingCompileTaskTest extends Specification {
 
         then:
         def e = thrown(Throwable)
-        def root = getRootCause(e)
+        def root = Throwables.getRootCause(e)
         root.message == errorMsg
-        that root.stackTrace, hasFrame(throwingClass)
+        root.stackTrace.any({ frame -> frame.className == throwingClass })
 
         where:
         processorClassName              | throwingClass                      | errorMsg
@@ -64,51 +57,14 @@ class AnnotationProcessingCompileTaskTest extends Specification {
     }
 
     def createProcessingTask(AnnotationProcessorDeclaration proc) {
-        Set<AnnotationProcessorDeclaration> procs = Collections.singletonList(proc)
-        return new AnnotationProcessingCompileTask(compileTask, procs, procPath, result) {
+        def noopClassloader = Stub(ClassLoader)
+        return new AnnotationProcessingCompileTask(Stub(JavaCompiler.CompilationTask), [proc] as Set<AnnotationProcessorDeclaration>, [], Stub(AnnotationProcessingResult)) {
             ClassLoader createProcessorClassLoader() {
-                return new LimitedClassLoader(noopLoader)
+                return new LimitedClassLoader(noopClassloader)
             }
         }
     }
 
-    def getRootCause(Throwable t) {
-        if (t.getCause() == null || t.getCause() == t) {
-            return t
-        }
-        return getRootCause(t.getCause())
-    }
-
-    private hasFrame(final String className) {
-        [
-            matches: { frames ->
-                def foundFrame = null
-
-                for (StackTraceElement frame : frames) {
-                    if (className.equals(frame.className)) {
-                        foundFrame = frame
-                    }
-                }
-
-                if (foundFrame == null) {
-                    return false
-                }
-
-                assert foundFrame.fileName != null
-                assert foundFrame.lineNumber > 0
-                assert foundFrame.methodName != null
-                return true
-            },
-
-            describeTo: { description ->
-                description.appendText("stack trace contain frame for class ${className}")
-            },
-
-            describeMismatch: { frames, description ->
-                description.appendValue(frames.toArrayString()).appendText(" did not have frame for class ${className}")
-            }
-        ] as BaseMatcher
-    }
 }
 
 class LimitedClassLoader extends ClassLoader {
@@ -134,12 +90,12 @@ class LimitedClassLoader extends ClassLoader {
 @SupportedAnnotationTypes(RunsFaultyProcessor.QUALIFIED_NAME)
 @SupportedSourceVersion(SourceVersion.RELEASE_6)
 class FaultyProcessor extends AbstractProcessor {
-    public FaultyProcessor() {
+    FaultyProcessor() {
         throw new RuntimeException("Whoops!")
     }
 
     @Override
-    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment env) {
+    boolean process(Set<? extends TypeElement> annotations, RoundEnvironment env) {
         return false
     }
 }

--- a/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
+++ b/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
@@ -25,13 +25,8 @@ import spock.lang.Specification
 
 import javax.annotation.processing.AbstractProcessor
 import javax.annotation.processing.RoundEnvironment
-import javax.annotation.processing.SupportedAnnotationTypes
-import javax.annotation.processing.SupportedSourceVersion
-import javax.lang.model.SourceVersion
 import javax.lang.model.element.TypeElement
 import javax.tools.JavaCompiler
-import java.lang.annotation.ElementType
-import java.lang.annotation.Target
 import java.lang.reflect.InvocationTargetException
 
 class AnnotationProcessingCompileTaskTest extends Specification {
@@ -84,13 +79,6 @@ class LimitedClassLoader extends ClassLoader {
     }
 }
 
-@Target(ElementType.TYPE)
-@interface RunsFaultyProcessor {
-    String QUALIFIED_NAME = "org.gradle.api.internal.tasks.compile.RunsFaultyProcessor"
-}
-
-@SupportedAnnotationTypes(RunsFaultyProcessor.QUALIFIED_NAME)
-@SupportedSourceVersion(SourceVersion.RELEASE_6)
 class FaultyProcessor extends AbstractProcessor {
     FaultyProcessor() {
         throw new RuntimeException("Whoops!")

--- a/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
+++ b/subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/AnnotationProcessingCompileTaskTest.groovy
@@ -32,6 +32,7 @@ import javax.lang.model.element.TypeElement
 import javax.tools.JavaCompiler
 import java.lang.annotation.ElementType
 import java.lang.annotation.Target
+import java.lang.reflect.InvocationTargetException
 
 class AnnotationProcessingCompileTaskTest extends Specification {
 
@@ -48,7 +49,8 @@ class AnnotationProcessingCompileTaskTest extends Specification {
         def e = thrown(Throwable)
         def root = Throwables.getRootCause(e)
         root.message == errorMsg
-        root.stackTrace.any({ frame -> frame.className == throwingClass })
+        root.stackTrace.any { frame -> frame.className == throwingClass }
+        !Throwables.getCausalChain(e).any { c -> c.class == InvocationTargetException.class }
 
         where:
         processorClassName              | throwingClass                      | errorMsg

